File: 프로그래머스/2/42626. 더 맵게/더 맵게.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    priority_queue<long long, vector<long long>, greater<long long>> q;
    for(int i = 0; i < scoville.size(); i++){
        q.push(scoville[i]);
    }
    
    int cnt = 0;
    while(q.top() < K && q.size() >= 2){
        int a = q.top();
        q.pop();
        int b = q.top();
        q.pop();
        
        long long tmp = (long long)a + ((long long)b * 2);
        q.push(tmp);
        cnt++;
    }
    if(q.top() < K)
        answer = -1;
    else
        answer = cnt;
    return answer;
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.java
import java.io.*;
import java.util.*;
import java.lang.*;

class Solution {
    PriorityQueue<Integer> pq = new PriorityQueue<>();
    
    public int solution(int[] scoville, int K) {
        int answer = 0;
        
        for(int i = 0; i < scoville.length; i++)
            pq.add(scoville[i]);
        
        if(pq.size() == 1){
            if(pq.peek() < K)
                answer = -1;
        }
        else{
            while(pq.size() != 1){
                int a = pq.peek();
                if(a >= K)
                    break;
                pq.poll();
                int b = pq.poll();
                
                pq.add(a + (b * 2));
                answer += 1;
            }
            if(pq.peek() < K)
                answer = -1;
        }
        return answer;
    }
}